### DISCUSSION TITLE
Ignore build step

### DISCUSSION THREAD

**Discussion by Neo-Oli** [vercel/vercel#5171] [Help] [answered]
Can you tell me what kind of commands are available in the ignore build step option? We want to filter deploys based on branch name.
```
# Branch name is 'feature/test-featurename'
echo "$VERCEL_GITHUB_COMMIT_REF"| grep -v test
```
But this will cause it to always skip building.
I tested it with the following two commands as well
```
sh -c 'exit 1'
sh -c 'exit 0'
```
But both cause the build to continue, so I am a bit confused.

**Comment by styfle** [comment]
Any command should work.

There is a similar discussion here: https://github.com/vercel/vercel/discussions/5142

**Reply by michaelbirchler** [reply]
@styfle thank you for your answer ans help. Could you make an example. I've pasted following code to _Ignored Build Step_
 field :
`echo $VERCEL_GITHUB_COMMIT_REF | grep -q notruf && exit 1` but the deployment always cancels.

I've tested my code locally and it works fine:
`echo feature/notruf-adjust-deployment | grep -q notruf || exit 1`

"notruf" is the string that we want to search in the current branch and if the string occurs, we want a deployment otherwise we want to cancel.

**Reply by svobik7** [reply]
**EDIT**: Sorry, I wrote it too fast. And now I see build was canceled. So it does not work for me either.

Hi @michaelbirchler, I have just successfully setup my new project on vercel with following command in _Ignored Build Step_:

`echo $VERCEL_GITHUB_COMMIT_REF | grep -q beta && exit 1 || exit 0`

So every push to branch without `beta` in its name is ignored to build.

**Reply by michaelbirchler** [reply]
@svobik7 thanks for testing 🙇‍♂️ . A working example or a better documentation from the Vercel team would be appreciated.

**Reply by cyrus-za** [reply]
The issue I'm experiencing in #5142 is that vercel seems to clone with depth=1 so no way to compare history to see what has changed. The response from vercel team seems to have gone quiet there so I figured it's worth checking other discussions for some help, but seems like these are also going unanswered

**Reply by michaelbirchler** [reply]
@cyrus-za @svobik7 I found an answer from @styfle here https://github.com/vercel/vercel/issues/3166#issuecomment-706225290
> Thanks, we'll get this fixed and document a proper solution!

**Reply by paulogdm** [reply]
We just released a guide about this field: https://vercel.com/knowledge/how-do-i-use-the-ignored-build-step-field-on-vercel

If you have any further suggestions, let us know.

**Reply by svobik7** [reply]
@michaelbirchler would you mind to share your settings here?

I am still not able to set ignored build correctly. What I am about to set is that if commit was pushed to branch `beta/dev` then process the build otherwise cancel any build. 

I tried following options inside Ignore Build Step Command: 
- `git branch --show-current | grep -q "beta" && exit 1 || exit 0`
- `git branch --show-current | grep -q "beta/*" && exit 1 || exit 0`
- `git branch --show-current | grep -q "beta/dev" && exit 1 || exit 0`

And also combination of `sh script.sh` while inside script.sh:

```
#!/bin/bash

# Retrieve branch name
branch=$(git branch --show-current)

if [ "$branch" == "beta/dev" ]; then
  # Proceed with build
    exit 1;
else
  # Don't build
  exit 0;
fi
```

Neither of these options work because any push to my repo invokes the deployment but after a few second the build process is cancelled:

```
15:01:39.937 | The Deployment has been canceled as a result of running the command defined in the "Ignored Build Step" setting.
15:01:40.413 | Deployment has been canceled
```

**Reply by michaelbirchler** [reply]
@svobik7 I didn't had a chance to try the proposed solution till tonight. I've tested it right now with the proposed script, but it always exits with 0. The script works locally just fine. I've debugged it locally with the proposed [debug instructions](https://vercel.com/knowledge/how-do-i-use-the-ignored-build-step-field-on-vercel#debugging-commands-locally).

@paulogdm do you have a clue why the following script doesn't work on the server:
```
#!/bin/bash

# Retrieve branch name
branch=$(git branch --show-current)

if [ "$branch" == "beta/dev" ]; then
  # Proceed with build
    exit 1;
else
  # Don't build
  exit 0;
fi
```
and you may share us some advice how to debug this? 🙇‍♂️

**Reply by cyrus-za** [reply]
I'm running into the exact same issue. My guess is that the exit code from the script is not persisted or something.

**Reply by paulogdm** [reply]
> I'm running into the exact same issue. My guess is that the exit code from the script is not persisted or something.

Can you elaborate?

**Reply by cyrus-za** [reply]
Elaboration is right above that and I tagged you in the other thread with more info on my situation and my script @paulogdm

**Reply by paulogdm** [reply]
I am not following: what "persisted" means in this context?

**Reply by Neo-Oli** [reply]
I did do some debugging of this, by writing to a file from within the ignore-build-step script.
`$VERCEL_GITHUB_COMMIT_REF` is empty at this point and `git rev-parse --abbrev-ref HEAD` always outputs `master` even though our repository doesn't have a master branch. `git branch --show-current` also outputs `master`.

**Reply by michaelbirchler** [reply]
@Neo-Oli thanks for [debugging](https://github.com/vercel/vercel/discussions/5171#discussioncomment-121417) the problem. @paulogdm would it be possible for you to investigate the problem? 🙇

**Reply by paulogdm** [reply]
Yes, I should investigate this in the next couple of days.

**Reply by paulogdm** [reply]
I think I've located the issue. Will give updates by the end of the week.

**Reply by hoomanaskari** [reply]
Just a quick question on this while you are working on the fix;

The Vercel documentation says:
"To run a bash script in the "Ignored Build Step", you need to set the following in the field: `sh script.sh`. Do notice the file should exist in your repository."

By `sh script.sh` you mean a file named `sh script.sh` (containing a space in the name) or a folder named `sh` with a file named `script.sh` inside the root directory of my react app? Thanks in advance.

**Reply by ys-chung** [reply]
@hoomanaskari A file named `script.sh` in the root of the repo.

**Reply by hoomanaskari** [reply]
Thank you very much. I can also confirm that the following code is not working for me:

```bash
#!/bin/bash

# Retrieve branch name
branch=$(git branch --show-current)

if [ "$branch" == "master" ]; then
  # Proceed with build
    exit 1;
else
  # Don't build
  exit 0;
fi
```

But I suppose a fix is on the way.

**Reply by paulogdm** [reply]
Hello everyone!

We've discussed this matter internally and we will release improvements to this command soon. For now, I will pull the guide from our knowledge DB. You can speak a big update before the end of the year.

**Reply by hunterwilhelm** [reply]
For anyone still trying to use `git branch --show-current` in the Ignore Build Step, it doesn't change between source branches. Instead, use `$VERCEL_GIT_COMMIT_REF`. [Read the tutorial](https://vercel.com/knowledge/how-do-i-use-the-ignored-build-step-field-on-vercel) from the answer & [All System Variables Documentation](https://vercel.com/docs/projects/environment-variables/system-environment-variables)
![Screenshot 2024-02-08 at 12 59 18 PM](https://github.com/vercel/vercel/assets/71348224/c311842c-7832-45d7-88cc-b27eb299bd56)

**Comment by paulogdm** [comment, marked as answer]
Hello!

With the release of our new Environment Variables UI, I have added a couple of examples to our KB article: https://vercel.com/knowledge/how-do-i-use-the-ignored-build-step-field-on-vercel

Feedbacks are welcome.

**Reply by paulogdm** [reply]
@hoomanaskari @ys-chung @michaelbirchler @cyrus-za @Neo-Oli

**Reply by dborstelmann** [reply]
I still have issues with this using any of the above.

```
branch=$(git rev-parse --abbrev-ref HEAD)
if [[ "$branch" == "master" || "$branch" == "test"  ]] ; then
...
```
^^^ Always builds even if not master or test (assuming the command always returns master)

```
if [[ "$VERCEL_GITHUB_COMMIT_REF" == "master" || "$VERCEL_GITHUB_COMMIT_REF" == "test"  ]] ; then
...
```
^^^ Never builds (assuming env variable is not set)

I use vercel for serverless connecting to an AWS database.  Because all commits to github push a new preview deployment (on ANY branch!!!) if anyone pushes it can push broken server code that corrupts the DB!  This is really frustrating.  And without ignored builds there is other way I know of to solve this.  So it would be great if we could get a solution here, and if not is there something I am missing on how to do this otherwise?

**Reply by hoomanaskari** [reply]
Still not working for me.

I enabled `Automatically expose System Environment Variables →` from my project settings, I also have a file named `script.sh` in root folder of my react app with the following code inside it:

```
#!/bin/bash

if [[ "$BRANCH" == "main" || "$BRANCH" == "master" ]] ; then
  # Proceed with the build
  exit 1;

else
  # Don't build
  exit 0;
fi
```

I pushed to a branch named new-canvas and vercel is still running the build.

**Reply by hoomanaskari** [reply]
I think Vercel should consider adding an option for this, Netlify has a built-in option that makes this very easy:

<img width="903" alt="Screen Shot 2020-11-25 at 13 21 57 PM" src="https://user-images.githubusercontent.com/69511785/100214902-75b2a400-2f21-11eb-8fd1-5cc29110226e.png">

**Reply by paulogdm** [reply]
@dborstelmann Please check your env variables.

`branch=$(git rev-parse --abbrev-ref HEAD)` will never work due to how we clone the repository.

**Reply by paulogdm** [reply]
@hoomanaskari That was an example. Did you add the environment variable?

![Screenshot from 2020-11-25 07-54-51](https://user-images.githubusercontent.com/7690649/100218503-982dc880-2ef3-11eb-8d2b-5014032f5fa9.png)

**Reply by svobik7** [reply]
@paulogdm your suggestion do not work for me either. Here is my configuration:

Variables config:
<img width="764" alt="Screenshot 2020-11-25 at 11 59 17" src="https://user-images.githubusercontent.com/761766/100219166-f7043980-2f15-11eb-8883-861793b4e8e2.png">

Ignore config:
<img width="785" alt="Screenshot 2020-11-25 at 11 59 35" src="https://user-images.githubusercontent.com/761766/100219172-f8356680-2f15-11eb-9ead-1eb726c8038c.png">

vercel.sh file:
<img width="425" alt="Screenshot 2020-11-25 at 12 01 58" src="https://user-images.githubusercontent.com/761766/100219262-0daa9080-2f16-11eb-87a6-8892debf719a.png">

results of deployments:
<img width="695" alt="Screenshot 2020-11-25 at 12 01 03" src="https://user-images.githubusercontent.com/761766/100219286-14d19e80-2f16-11eb-8a95-b952f7275226.png">

Is there anything I missed?

**Reply by Neo-Oli** [reply]
I don't think environment variables get exposed to the ignore-build-step script at all. I tested it with this script:
```
#!/bin/sh
echo "VERCEL_GITHUB_COMMIT_REF = $VERCEL_GITHUB_COMMIT_REF" >> public/static/env
echo "BRANCH = $BRANCH" >> public/static/env
echo "GOOGLE_TAG_MANAGER_ID = $GOOGLE_TAG_MANAGER_ID" >> public/static/env
exit 1 #always build
```
but the resulting file is empty:
```
VERCEL_GITHUB_COMMIT_REF =
BRANCH =
GOOGLE_TAG_MANAGER_ID =
```

![Screenshot_2020-11-25 Environment Variables - notruf-redcross-ch-v2 – Vercel](https://user-images.githubusercontent.com/7983745/100224770-8c56fc00-2f1d-11eb-88ae-f0f64ebacd06.png)

**Reply by paulogdm** [reply]
Allow me to do a public deployment. I will get back to you in a bit.

**Reply by jamespb97** [reply]
@Neo-Oli Not sure if it makes any difference but have you tried `VERCEL_GIT_COMMIT_REF` instead of `VERCEL_GITHUB_COMMIT_REF`?
<img width="742" alt="Screenshot 2020-11-25 at 12 20 13" src="https://user-images.githubusercontent.com/23367467/100227250-c40f7500-2f18-11eb-9afa-c1d7d16543b2.png">

**Reply by paulogdm** [reply]
Hey everyone,

My sincere apologies. We have finally located the issue! A fix is being released as we speak.

**Reply by Neo-Oli** [reply]
@jamespb97 Thanks, i just tried these, but these are also empty.

**Reply by hoomanaskari** [reply]
in my case all builds fail. Investigating right now to make sure if the issue with my code or Vercel. Will report back.

**Reply by hoomanaskari** [reply]
It appears to be caused by Vercel. After disabling `Automatically expose System Environment Variables` and removing `bash script.sh` it started working again!

**Reply by paulogdm** [reply]
Don't worry, a PR is on the way to fix this once and for all.

**Reply by jamespb97** [reply]
@paulogdm Is there any update on this or ETA on when we can expect a fix to be deployed? Thanks :)

**Reply by paulogdm** [reply]
We got side-tracked by the `us-east-1` outage. I will get back to you today.

**Reply by paulogdm** [reply]
We have confirmed the fix. It will go live to all accounts in a few hours.

**Reply by paulogdm** [reply]
This is now live for all regions & accounts. Sorry, everyone, for the delay!

**Reply by svobik7** [reply]
@paulogdm thank you, it is working now.

**Reply by dborstelmann** [reply]
@paulogdm Working for me as well, thanks for the quick turnaround.

**Reply by bfabio** [reply]
@paulogdm We are using the external script approach, but we want to skip the `gh-pages` branch, which is automatically pushed with only the artifacts of the build. Is there a way to do it without pushing the script itself to the `gh-pages` branch?

Being able to use shell piping in the command would fix our issues.

**Reply by jonashaefele** [reply]
Related to the guide @paulogdm linked to more than the discussion above. Just found this after getting a bit stuck...

I set up two vercel projects from a mono-repo sort of like described in https://vercel.com/docs/git?query=monorepo

My structure is like this:

```
/shared
    /components
    /lib
    /...
/app-one
/app-two
```

`app-one` and `app-two` share a design system and a few config files from `shared` linked with yarn workspaces and a little tsconfig magic. That all works perfectly.

But it took me ages to realise that the ignore build step command is running in the root folder of the individual deployment, and not the root of the repo. It's all linked and discoverable. So all totally my fault.

But might be worth adding an example at https://vercel.com/docs/git?query=monorepo#ignoring-the-build-step or in the guide mentioned above at https://vercel.com/knowledge/how-do-i-use-the-ignored-build-step-field-on-vercel#with-folders-and-workspaces

Ignore commands in my case were:

For `app-one`: `git diff --quiet HEAD^..HEAD -- .. ':!app-two'`
For `app-two`: `git diff --quiet HEAD^..HEAD -- .. ':!app-one'`

It's literally just those two little dots that made it work. Took me a while to get that I need to go one level up, from the app root to the repo root. So it's `-- .. ':!app-two'` not `-- . ':!app-two'`

This way both changes to the `shared` design system and the app that's getting deployed let the deployment continue, but changes to the unrelated app abort the deployment.

**Reply by gregg-cbs** [reply]
Oh my goodness! I gave up on this and then about 3 months later started looking into it... the `..` is the saving grace.

Why on earth vercel do you set the default path for this as the root directory. We are talking about repository level routes here and you start it inside the project root. Im sure many people dont understand this.

You need to update your docs to make this clear because without Neo i would never have known.
So essentially most people would want `./`

**Comment by maltesa** [comment]
I'm trying to make this work, as I want to deploy when there are changes between the last deployment and the current HEAD:
```
git diff $VERCEL_GIT_PREVIOUS_SHA HEAD --quiet ./app1
```

But that fails with a git error (runs fine on my local machine, though):
`fatal: bad object b0eae4da9dc8c129fdda5cb06d2bf616f3b6c623` exit code 128

I don't quite get why the following makes sense. What happens if I push multiple commits, but the last one did not change anything in `./app1`? In that case, no build would be triggered, right?
```
git diff HEAD^ HEAD --quiet ./app1
```
